src/reason_pprint_ast.ml

```diff
@@ -482,133 +527,239 @@ let rec sequentialIfBlocks x =
+   considered a "simple expression" (it's acceptable anywhere a simple
+   expression is except in a couple of edge cases.
+
+   "Unary Prefix" function applications parse with the rule:
+
+   `MINUS epxr %prec prec_unary_minus`, which in turn is considered an
+   "expression" (not simple). All unary operators are mapped into an identifier
+   beginning with "~".
+
+   TODO: Migrate all "almost simple prefix" to "unsary prefix". When `!`
+   becomes "not", then it will make more sense that !myFunc (arg) is parsed as
+   !(myFunc arg) instead of (!myFunc) arg.
 
-let prefix_symbols  = [ '!'; '?'; '~' ] ;;
+ *)
+let almost_simple_prefix_symbols  = [ '!'; '?'; '~'] ;;
+(* Subset of prefix symbols that have special "unary precedence" *)
+let unary_minus_prefix_symbols  = [ "~-"; "~-."] ;;
+let unary_plus_prefix_symbols  = ["~+"; "~+." ] ;;
 let infix_symbols = [ '='; '<'; '>'; '@'; '^'; '|'; '&'; '+'; '-'; '*'; '/';
                       '$'; '%'; '\\' ]
+(* There are just a couple of infix operators that start with `!`. All others that
+   start with ! would be prefix, so we keep a list of exceptions here. *)
+let infix_starting_with_bang = ["!="; "!=="]
 let operator_chars = [ '!'; '$'; '%'; '&'; '*'; '+'; '-'; '.'; '/';
                        ':'; '<'; '='; '>'; '?'; '@'; '^'; '|'; '~' ]
 let numeric_chars  = [ '0'; '1'; '2'; '3'; '4'; '5'; '6'; '7'; '8'; '9' ]
 
-let indexOfFirstMatch s lst =
-  let rec indexOfFirstMatchN s lst n = match lst with
-    | [] -> None
-    | []::tl -> indexOfFirstMatchN s tl (n + 1)
-    | (hdHd::hdTl)::tl -> (
-        match hdHd s with
-          | (prec, true) -> Some (prec, n)
-          | (_, false) -> indexOfFirstMatchN s (hdTl::tl) (n)
-    )
-  in
-  indexOfFirstMatchN s lst 0
+let special_infix_strings =
+  ["asr"; "land"; "lor"; "lsl"; "lsr"; "lxor"; "mod"; "or"; ":="; "!="; "!=="]
 
+let updateToken = "="
+let requireIndentFor = [updateToken; ":="]
+
+let infixTokenRequiresIndent printedIdent =
+  if List.exists (fun i -> i = printedIdent) requireIndentFor then None else Some 0
+
+let getPrintableUnaryIdent s =
+  if List.mem s unary_minus_prefix_symbols || List.mem s unary_plus_prefix_symbols then
+    String.sub s 1 (String.length s -1)
+  else s
+
+
+(* determines if the string is an infix string.
+   checks backwards, first allowing a renaming postfix ("_102") which
+   may have resulted from Pexp -> Texp -> Pexp translation, then checking
+   if all the characters in the beginning of the string are valid infix
+   characters. *)
+let printedStringAndFixity  = function
+  | s when List.mem s special_infix_strings -> Infix s
+  | s when List.mem s.[0] infix_symbols || List.mem s infix_starting_with_bang -> Infix s
+  (* Correctness under assumption that unary operators are stored in AST with
+     leading "~" *)
+  | s when List.mem s.[0] almost_simple_prefix_symbols &&
+           not (List.mem s infix_starting_with_bang) &&
```

Comment:
Is it redundant? It seems to describe the exact opposite. I thought this was good because it makes each individual match case truthful on its own, and the entire set capable of being reordered without changing the meaning of the entire pattern match.